api/msam/chalicelib/cloudwatch.py

```diff
@@ -134,14 +134,43 @@ def get_cloudwatch_events_state(state):
     """
     API entry point to retrieve all pipeline events in a given state (set, clear).
     """
+    events = []
     dynamodb = boto3.resource('dynamodb')
     table = dynamodb.Table(EVENTS_TABLE_NAME)
     response = table.query(IndexName='AlarmStateIndex', KeyConditionExpression=Key('alarm_state').eq(state))
     if "Items" in response:
-        alarms = response["Items"]
-    else:
-        alarms = []
-    return alarms
+        events = response["Items"]
+    return events
+
+
+def get_cloudwatch_events_state_groups(state):
+    """
+    Group all events by down, degraded and running pipelines
+    """
+    group = {}
+    group["down"] = []
+    group["running"] = []
+    group["degraded"] = []
+    events = get_cloudwatch_events_state(state)
+    for event in events:
+        resource_arn = event["resource_arn"]
+        def is_same_arn(i):
+            return bool(i["resource_arn"] == resource_arn)
+        def is_pl_down(i):
+            return bool("pipeline_state" in i["detail"] and not i["detail"]["pipeline_state"])
+        same_events = list(filter(is_same_arn, events))
```

Comment:
Even though we've already grouped events of the same arn, those events will still be in the events list the next time you iterate over so we could be processing same arn more than once.

Reply:
Not necessarily. 
This for-loop essentially runs once per event in the array of events. For each event we do the following:

- `same_events` - get all events in the events array that have the same arn, including the current event in question.
- `down_pipelines` - from the `same_events` results, get all events with down pipelines.
  - if the count of `down_pipelines` is 1 (including the event in question), were degraded.
    - add current event in question to degraded array
  - if the count of `down_pipelines` is more than 1 (including the event in question), were down.
    - add current event in question to down array
  - if the count of `down_pipelines` is 0 (including the event in question), were running.
    - add current event in question to running array

IOW, I just want to know for the event in question, how many pipelines are down including the event in question. Uhm, I hope that makes sense.